src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -307,6 +278,20 @@ def __init__(
             logger.debug("Using cached wheel link: %s", cache_entry.link)
             link = cache_entry.link
         ireq = make_install_req_from_link(link, template)
+        assert ireq.link == link
+        if ireq.link.is_wheel and not ireq.link.is_file:
+            wheel = Wheel(ireq.link.filename)
+            wheel_name = canonicalize_name(wheel.name)
+            assert name == wheel_name, (
+               "{!r} != {!r} for wheel".format(name, wheel_name)
+            )
+            # Version may not be present for PEP 508 direct URLs
+            if version is not None:
+                assert str(version) == wheel.version, (
```

Comment:
I'm wondering about these checks: what is the difference from _check_metadata_consistency?  I initially thought that these are to fail early, but assertions are not executed in compiled code.

Reply:
We use assertions to:

1. express assumptions that we're operating under, and that the code may depend on
2. enforce invariants in objects
3. generally to guard against developer errors (as opposed to user errors, which should raise)

In this case, I switched to extracting the name and version from the link in 9e46391, but didn't feel like ignoring the existing name and version was the right thing to do. Because of the way we do processing the assertions are true now, but that may change later and we'd want to catch it if it does.

As far as I know assertions are only disabled when optimization is turned on, so I'd expect in the majority of cases (like our tests and anyone using default Python options) the assertions will execute.

Reply:
Thank you for the insight, I think I've just learnt something new today.